app/com/arpnetworking/metrics/portal/reports/ReportExecution.java

```diff
@@ -0,0 +1,184 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.arpnetworking.metrics.portal.reports;
+
+import com.google.common.base.CaseFormat;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSetMultimap;
+import com.google.common.collect.Maps;
+import com.google.inject.Injector;
+import com.google.inject.Key;
+import com.google.inject.name.Names;
+import models.internal.impl.ChromeScreenshotReportSource;
+import models.internal.impl.HtmlReportFormat;
+import models.internal.impl.PdfReportFormat;
+import models.internal.reports.Recipient;
+import models.internal.reports.Report;
+import models.internal.reports.ReportFormat;
+import models.internal.reports.ReportSource;
+
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+
+/**
+ * Utilities for execution of {@link Report}s.
+ *
+ * @author Spencer Pearson (spencerpearson at dropbox dot com)
+ */
+public final class ReportExecution {
+
+    /**
+     * Render and send a report.
+     *
+     * @param report The report to render and execute.
+     * @param injector A Guice injector to pull dependencies out of.
+     * @param scheduled The instant the report was scheduled for.
+     * @return A CompletionStage that completes when sending has completed for every recipient.
+     */
+    public static CompletionStage<Report.Result> execute(final Report report, final Injector injector, final Instant scheduled) {
+        verifyDependencies(report, injector);
+        final ImmutableMultimap<ReportFormat, Recipient> formatToRecipients = report.getRecipientsByFormat()
+                .entrySet()
+                .stream()
+                .collect(ImmutableSetMultimap.flatteningToImmutableSetMultimap(
+                        Map.Entry::getKey,
+                        e -> e.getValue().stream()));
+        final ImmutableMultimap<Recipient, ReportFormat> recipientToFormats = formatToRecipients.inverse();
+        return renderAll(injector, formatToRecipients.keySet(), report.getSource(), scheduled)
+                .thenCompose(
+                        formatToRendered -> sendAll(injector, recipientToFormats, formatToRendered)
+                ).thenApply(
+                        nothing -> null
```

Comment:
Can you elaborate on what this is doing?

Reply:
As discussed if this is temporary and if it's easy let's just type the CompletionStage to `Void` for now.